src/payment.rs

```diff
@@ -52,6 +53,32 @@ pub fn may_pay(info: &MessageInfo, denom: &str) -> Result<Uint128, PaymentError>
     }
 }
 
+fn stringify_coins(coins: &[Coin]) -> String {
+    match coins {
+        [] => "None".to_string(),
+        _ => coins
+            .iter()
+            .map(ToString::to_string)
+            .collect::<Vec<_>>()
+            .join(", "),
+    }
+}
```

Comment:
It looks off - it should happen only when the error is printed. If you really want to have control about formatting, just wrap the `Vec<Coin>` in the error type in the newtype.

Reply:
Not sure what you mean. This function currently does only execute on the error case. The intention is to provide a clear error to the consumer.

> just wrap the Vec<Coin> in the error type in the newtype

Do you have an example? Are you saying to create a new type that wraps Vec<Coin> that implements the Display trait?

All said, I think the best approach is still to implement the Display trait in cosmwasm-std.